### PR TITLE
feat: Support for iOS & macOS using Cocoa SDK integration

### PR DESCRIPTION
**Pull request by limbonaut**
This PR brings Sentry Cocoa SDK integration as a backend on macOS and iOS platforms. This switches macOS builds to utilize Sentry Cocoa as well.

Build system manages Sentry Cocoa package, reading `modules/sentry-cocoa.package` file, downloading the corresponding package with dynamic framework from GitHub Releases, extracting and processing it. It also updates it if the version changes in the .package file. Dependency update should be trivial to add.

The build system takes the respective slices from the Sentry Cocoa XCFramework and puts them in the `project/`. For macOS, it also flattens the framework, so we don’t have any problems with [symlinks](https://github.com/getsentry/sentry-cocoa/issues/5740) on Windows and with Godot Engine. Framework is linked dynamically.

This PR also includes a simple system for custom options. GodotCPP tends to complain about those – so there is a workaround for that too. A new option is added to generate iOS framework from iOS build artifacts: `generate_ios_framework=yes|no`

<details>
<summary>Local build instructions</summary>

Build macOS libs:
```
scons target=editor debug_symbols=yes
scons target=template_release debug_symbols=yes 
```

Build iOS libs:
```
scons platform=ios target=editor arch=arm64 ios_simulator=no
scons platform=ios target=editor arch=universal ios_simulator=yes generate_ios_framework=yes
```

</details>

All unit tests passed on macOS. Due to limitations of the unit test framework, can't run tests on iOS. However, there are separate mobile tests which I ran via demo project – all succeeded.

### Known issues and limitations

1. Script source context is not available.
2. Local variables are not available.
3. Scene tree information is not available (aka view hierarchy).
4. CI doesn't make proper iOS builds yet. It does compilation, but does not generate XCFramework yet.
5. Debug output is not routed to Godot console yet. You can see it in a terminal app. It's also quite noisy.
6. Godot generates runtime errors when `OS.get_memory_info()` is called on iOS – so I disabled it on iOS.

### CODE REVIEW

File: src/sentry/cocoa/cocoa_sdk.mm [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+			if (unlikely(processed.is_null())) {
+				return nil;
+			} else {
+				last_event_id_mutex->lock();
+				last_event_id = string_from_objc(event.eventId.sentryIdString);
+				last_event_id_mutex->unlock();
+				return event;
+			}
+		};
+	}];
```
[the remark this answers is not in the document]
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_sdk.mm:200]: 1. process_event() cant return a different object: it returns the same object or null.
2. exceptions are disabled.

**Comment by bruno-garcia** [comment]
bugbot review
@sentry review

File: src/sentry/cocoa/cocoa_sdk.mm [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+				//       Threads override our custom-built stacktrace for scripts.
+				// TODO: Try adding script stack trace in "threads" instead.
+				event.threads = nil;
+				event.debugMeta = nil;
+			}
+
+			Ref<CocoaEvent> event_obj = memnew(CocoaEvent(event));
+			Ref<CocoaEvent> processed = sentry::process_event(event_obj);
+
+			if (unlikely(processed.is_null())) {
```
[the remark this answers is not in the document]
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_sdk.mm]: exceptions are disabled

File: src/sentry/util/timestamp.cpp [hunk 3]
```diff
@@ -1,10 +1,149 @@
...
+
+	int year, month, day, hour, minute, second = 0;
+	int64_t nanos = 0;
+
+	int num_consumed = 0;
+	constexpr int num_inputs = 6;
+	constexpr int num_chars_in_date = 19;
+	if (sscanf(cur, "%d-%d-%dT%d:%d:%d%n", &year, &month, &day, &hour, &minute, &second, &num_consumed) < num_inputs || num_consumed != num_chars_in_date) {
+		return 0;
+	}
```
[the remark this answers is not in the document]
Comment (limbonaut) [reply src/sentry/util/timestamp.cpp]: Resolved in 1bbe21e90258a330cc28849c09e1413bf758b8c0

File: src/sentry/util/timestamp.cpp [hunk 4]
```diff
@@ -1,10 +1,149 @@
...
+		if (sscanf(cur, ".%d%n", &fractional, &num_consumed) < 1 || num_consumed > 10) {
+			sentry::util::print_error("Timestamp parsing needs 1-9 fractional digits.");
+			return 0;
+		}
+		cur += num_consumed;
+
+		// normalize to nanoseconds
+		int missing_digits = 10 - num_consumed;
+		while (missing_digits--) {
+			fractional *= 10;
```
[the remark this answers is not in the document]
Comment (limbonaut) [reply src/sentry/util/timestamp.cpp]: there is validation in sscanf above

File: src/sentry/cocoa/cocoa_util.mm [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+			const Array &keys = dict.keys();
+			for (int i = 0; i < keys.size(); i++) {
+				const String &key = keys[i];
+				const NSString *objc_key = [NSString stringWithUTF8String:key.utf8()];
+				const NSObject *objc_value = variant_to_objc(dict[key]);
+				[objc_dict setObject:objc_value forKey:objc_key];
+			}
+
+			return objc_dict;
```
[the remark this answers is not in the document]
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_util.mm]: Addressed in 506561602c08b44dcfbd656d794e283783b407e6

File: src/sentry/cocoa/cocoa_util.mm [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		case Variant::PACKED_VECTOR4_ARRAY: {
+			NSMutableArray *objc_array = [[NSMutableArray alloc] init];
+			bool oob = false;
+			bool valid = true;
+			int i = 0;
+
+			do {
+				Variant item = p_value.get_indexed(i++, valid, oob);
+				if (valid) {
+					[objc_array addObject:variant_to_objc(item)];
```
[the remark this answers is not in the document]
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_util.mm]: Fixed in 506561602c08b44dcfbd656d794e283783b407e6

**Comment by jpnurmi** [comment]
Could we have some basic unit tests for `rfc3339_timestamp_to_microseconds` and `microseconds_to_rfc3339_timestamp`? A separate PR could also be a welcome option to offload some parts of this rather heavy PR and get things going. :)

**Comment by limbonaut** [comment]
@jpnurmi The tests we maintain are in GDScript, which don't have direct access to such methods; however, we test them indirectly via: https://github.com/getsentry/sentry-godot/blob/efbdb39a5c43b8fc754370c711bd62cb5a96c202/project/test/suites/test_event.gd#L28-L45

We could expand such tests to include various formats in RFC3339 and also test some edge cases. Extracting these functions into separate PR make less sense as they only used with cocoa sdk, and we wouldn't be able to test them unless we introduce another testing layer, just for internal C++ stuff.

**Comment by limbonaut** [comment]
@jpnurmi Almost forgot, I’ve been working on a dedicated `SentryTimestamp` public API class [some time ago](https://github.com/getsentry/sentry-godot/compare/main...feat/timestamp-class). I stopped working on it, as I’m not entirely sure if we really need it, so I’d love to hear your thoughts. It actually has a pretty comprehensive test suite :)
One of the limitations it tries to overcome is lack of dedicated datetime/timestamp type in GDScript. This is all we have in Godot: https://docs.godotengine.org/en/latest/classes/class_time.html

UPDATE:
- See #286

**Comment by limbonaut** [comment]
Timestamp functionality is taken care of by #286. So, this PR doesn’t deal with it anymore.

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,7 +4,8 @@
 
 ### Features
 
-- Intoduce `SentryTimestamp` class to improve timestamp handling ([#286](https://github.com/getsentry/sentry-godot/pull/286))
+- Intoduce `SentryTimestamp` class to improve timestamp handling ([#286](https://github.com/getsentry/sentry-godot/pull/286
```
Comment (jpnurmi) [inline CHANGELOG.md]: trailing `))` went missing in merge

Comment (limbonaut) [reply CHANGELOG.md]: O.o  nice catch!

Comment (limbonaut) [reply CHANGELOG.md]: Fixed in 19ac7505ee9e7cb1e8c6bbe4ffb288f0d2d6d67c

File: src/sentry/cocoa/cocoa_util.mm [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+NSObject *variant_to_objc(const godot::Variant &p_value, int p_depth) {
+	switch (p_value.get_type()) {
+		case Variant::NIL: {
+			return [NSNull null];
+		}
+		case Variant::BOOL: {
+			return [NSNumber numberWithBool:(bool)p_value];
+		}
+		case Variant::INT: {
+			return [NSNumber numberWithLong:(int64_t)p_value];
```
Comment (jpnurmi) [inline src/sentry/cocoa/cocoa_util.mm]: should it be `numberWithLongLong`?

File: src/sentry/cocoa/cocoa_util.h [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+_FORCE_INLINE_ NSString *string_to_objc_or_nil_if_empty(const godot::String &p_str) {
+	return p_str.is_empty() ? nil : [NSString stringWithUTF8String:p_str.utf8()];
+}
+
+_FORCE_INLINE_ godot::String string_from_objc(NSString *p_str) {
+	return p_str ? godot::String::utf8([p_str UTF8String]) : godot::String();
+}
+
+_FORCE_INLINE_ NSNumber *int_to_objc(int p_num) {
+	return [NSNumber numberWithLong:p_num];
```
Comment (jpnurmi) [inline src/sentry/cocoa/cocoa_util.h]: why `numberWithLong` instead of `numberWithInt`?

File: src/sentry/cocoa/cocoa_event.mm [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+	ERR_FAIL_NULL_V(cocoa_event, Ref<SentryTimestamp>());
+
+	if (cocoa_event.timestamp == nil) {
+		return Ref<SentryTimestamp>();
+	}
+
+	NSTimeInterval seconds = [cocoa_event.timestamp timeIntervalSince1970];
+	int64_t microseconds = (int64_t)(seconds * 1000'000.0);
+
+	return SentryTimestamp::from_microseconds_since_unix_epoch(microseconds);
```
Comment (jpnurmi) [inline src/sentry/cocoa/cocoa_event.mm]: wasn't there something like `from_unix_time` that takes seconds since epoch?

File: src/sentry/cocoa/cocoa_sdk.mm [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+	breadcrumb.level = sentry_level_to_objc(p_level);
+	if (!p_message.is_empty()) {
+		breadcrumb.message = string_to_objc(p_message);
+	}
+	if (!p_category.is_empty()) {
+		breadcrumb.category = string_to_objc(p_category);
+	}
+	if (!p_type.is_empty()) {
+		breadcrumb.type = string_to_objc(p_type);
+	}
```
Comment (jpnurmi) [inline src/sentry/cocoa/cocoa_sdk.mm]: `string_to_objc_or_nil_if_empty`?

File: src/sentry/cocoa/cocoa_sdk.mm [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+		}
+		if (!p_user->get_username().is_empty()) {
+			user.username = string_to_objc(p_user->get_username());
+		}
+		if (!p_user->get_email().is_empty()) {
+			user.email = string_to_objc(p_user->get_email());
+		}
+		if (!p_user->get_ip_address().is_empty()) {
+			user.ipAddress = string_to_objc(p_user->get_ip_address());
+		}
```
Comment (jpnurmi) [inline src/sentry/cocoa/cocoa_sdk.mm]: would `string_to_objc_or_nil_if_empty` be convenient here?

File: src/sentry/cocoa/cocoa_event.mm [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+
+	return cocoa_event.error != nil;
+}
+
+CocoaEvent::CocoaEvent() {
+	cocoa_event = [[objc::SentryEvent alloc] init];
+}
+
+CocoaEvent::CocoaEvent(objc::SentryEvent *p_cocoa_event) {
+	cocoa_event = p_cocoa_event;
```
Comment (jpnurmi) [inline src/sentry/cocoa/cocoa_event.mm]: nit: could be done in the initializer list

**Review by jpnurmi — COMMENTED** [review]
Looks good! Thanks for splitting off the timestamp changes. 🙏

File: src/sentry/cocoa/cocoa_event.mm (hunk 10 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_event.mm]: Yep, thanks!

File: src/sentry/cocoa/cocoa_util.mm (hunk 8 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_util.mm]: Resolved in c1aa27deb62eb9bbae8a9002167361db39e78ce6

File: src/sentry/cocoa/cocoa_util.h (hunk 9 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_util.h]: Resolved in c1aa27deb62eb9bbae8a9002167361db39e78ce6

File: src/sentry/cocoa/cocoa_event.mm (hunk 10 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_event.mm]: Resolved in a9c4bfd7d478d769fa3d9cc48c94e40b81950cab

File: src/sentry/cocoa/cocoa_sdk.mm (hunk 11 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_sdk.mm]: Resolved in b5728c79a1f422aab2ca4601a1a7d497b83e54b6

File: src/sentry/cocoa/cocoa_sdk.mm (hunk 12 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_sdk.mm]: Resolved b5728c79a1f422aab2ca4601a1a7d497b83e54b6

File: src/sentry/cocoa/cocoa_event.mm (hunk 13 above)
Comment (limbonaut) [reply src/sentry/cocoa/cocoa_event.mm]: Addressed in 2602f31b5f2f0bbb8fdee263658fe72b1b7e036f

**Review by jpnurmi — APPROVED** [review]
Are these "noop" builds anymore?

**Comment by jpnurmi** [comment]
Oh no, switching to Cocoa didn't help with that :slightly_frowning_face: 

> libc++abi: Pure virtual function called!
> /Users/runner/work/_temp/f74de61c-e6cf-42b0-8d86-f03e8521ab5d.sh: line 3:  1754 Abort trap: 6           ${GODOT} --headless --path project/ --import --verbose
> Error: Process completed with exit code 134.

https://github.com/getsentry/sentry-godot/actions/runs/16770287480/job/47484472250?pr=266

**Comment by limbonaut** [comment]
> Oh no, switching to Cocoa didn't help with that 🙁

Aww shame, cant even pin it to a commit as it's so flaky.

**Comment by limbonaut** [comment]
> Are these "noop" builds anymore?

They are not, but I'd like to address it in a separate PR as it needs more than renaming – we also need to generate an xcframework from 2 libs in a post-process job.
